pubsub/pubsub.go

```diff
@@ -404,37 +449,22 @@ func (s *Subscription) Receive(ctx context.Context) (_ *Message, err error) {
 		}
 		// No messages are available and there are no calls to ReceiveBatch in flight.
 		// Make a call.
-		//
-		// Ask for a number of messages that will give us the desired queue length.
-		// Unless we don't have information about process time (at the beginning), in
-		// which case just get one message.
-		nMessages := 1
-		if s.dynamicBatchSizes && s.avgProcessTime > 0 {
-			// Using Ceil guarantees at least one message.
-			n := math.Ceil(desiredQueueDuration.Seconds() / s.avgProcessTime)
-			// Cap nMessages at some non-ridiculous value.
-			// Slight hack: we should be using a larger cap, like MaxInt32. But
-			// that messes up replay: since the tests take very little time to ack,
-			// n is very large, and since our averaging process is time-sensitive,
-			// values can differ slightly from run to run. The current cap happens
-			// to work, but we should come up with a more robust solution.
-			// (Currently it doesn't matter for performance, because gcppubsub
-			// caps maxMessages to 1000 anyway.)
-			nMessages = int(math.Min(n, 1000))
-		}
+		batchSize := s.updateBatchSize()
 		s.waitc = make(chan struct{})
 		s.mu.Unlock()
 		// Even though the mutex is unlocked, only one goroutine can be here.
 		// The only way here is if s.waitc was nil. This goroutine just set
 		// s.waitc to non-nil while holding the lock.
 		if s.preReceiveBatchHook != nil {
-			s.preReceiveBatchHook(nMessages)
+			s.preReceiveBatchHook(batchSize)
 		}
-		msgs, err := s.getNextBatch(ctx, nMessages)
+		msgs, err := s.getNextBatch(ctx, batchSize)
 		if s.postReceiveBatchHook != nil {
 			s.postReceiveBatchHook(len(msgs))
 		}
 		s.mu.Lock()
+		s.lastBatchRecv = time.Now()
```

Comment:
I feel like you shouldn't update these values until you know err == nil.  Maybe it doesn't matter?

Reply:
Hm, interesting. Let's see.

If we update them after an error, the next call to `Receive` will try again to get more messages, and update the batch size based on 0 messages received over some small period of time, so the batch size will likely go down.

Same thing if this `ReceiveBatch` didn't return an error, but returned 0 messages.

If we don't update after an error, the next call to `Receive` will update the batch size based the number of messages received on the last successful `ReceiveBatch`, but over a longer period of time, so batch size will also go down, but probably by less.

So, I don't think it matters too much; always updating feels more correct to me since otherwise we're re-using the same values to adjust the batch size multiple times.

Added a comment.

Reply:
BTW, may have to reconsider this when we start calling `ReceiveBatch` before there are 0 messages left.